src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs

```diff
@@ -0,0 +1,329 @@
+            Host.CheckUserArg(args.Eta > 0, nameof(args.Eta), posError);
+
+            _lambda = args.Lambda;
+            _k = args.K;
+            _iter = args.NumIterations;
+            _eta = args.Eta;
+            _threads = args.NumThreads ?? Environment.ProcessorCount;
+            _quiet = args.Quiet;
+            _doNmf = args.NonNegative;
+
+            Info = new TrainerInfo(normalization: false, caching: false);
+        }
+
+        /// <summary>
+        /// Initializing a new instance of <see cref="MatrixFactorizationTrainer"/>.
+        /// </summary>
+        /// <param name="env">The private instance of <see cref="IHostEnvironment"/>.</param>
+        /// <param name="labelColumn">The name of the label column.</param>
+        /// <param name="xColumnName">The name of the column hosting the matrix's column IDs.</param>
+        /// <param name="yColumnName">The name of the column hosting the matrix's row IDs.</param>
+        /// <param name="advancedSettings">A delegate to apply all the advanced arguments to the algorithm.</param>
+        /// <param name="context">The <see cref="TrainerEstimatorContext"/> for additional input data to training.</param>
+        public MatrixFactorizationTrainer(IHostEnvironment env, string labelColumn, string xColumnName, string yColumnName,
+            TrainerEstimatorContext context = null, Action<Arguments> advancedSettings = null)
+            : base(env, LoadNameValue)
+        {
+            var args = new Arguments();
+            advancedSettings?.Invoke(args);
+
+            _lambda = args.Lambda;
+            _k = args.K;
+            _iter = args.NumIterations;
+            _eta = args.Eta;
+            _threads = args.NumThreads ?? Environment.ProcessorCount;
+            _quiet = args.Quiet;
+            _doNmf = args.NonNegative;
+
+            Info = new TrainerInfo(normalization: false, caching: false);
+            Context = context;
+
+            LabelColumn = new SchemaShape.Column(labelColumn, SchemaShape.Column.VectorKind.Scalar, NumberType.R4, false);
+            XColumn = new SchemaShape.Column(xColumnName, SchemaShape.Column.VectorKind.Scalar, NumberType.U4, true);
+            YColumn = new SchemaShape.Column(yColumnName, SchemaShape.Column.VectorKind.Scalar, NumberType.U4, true);
+        }
+
+        /// <summary>
+        /// Train a matrix factorization model based on training data, validation data, and so on in the given context.
+        /// </summary>
+        /// <param name="context">The information collection needed for training. <see cref="TrainContext"/> for details.</param>
+        public override MatrixFactorizationPredictor Train(TrainContext context)
+        {
+            Host.CheckValue(context, nameof(context));
+
+            using (var ch = Host.Start("Training"))
+            {
+                var pred = TrainCore(ch, context.TrainingSet, context.ValidationSet);
+                return pred;
```

Comment:
just return TrainCore, what's the point of pred?

Reply:
Fixed. Thanks.